### PR TITLE
Add `shadow`, `lineposition` and `textcase` options to SVG fonts

### LINKED ISSUE #6974: Support `text-shadow` in SVG text
https://developer.mozilla.org/en-US/docs/Web/CSS/text-shadow

We use these property in few places e.g. the `parcoords` trace.
It would be nice to have it for other text to help improve readability (cc: #4674) and more.

### PR DESCRIPTION
**Pull request by archmoj**
Resolves #6974 also helps address #4674.
Support additional text font styling options based on [`text-shadow`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-shadow), [`text-decoration-line`](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/text-decoration) & [`text-transform`](https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform).

Please see `src/plots/font_attributes.js` for these new optoins.

@plotly/plotly_js 
cc: @ndrezn

### CODE REVIEW

**Comment by archmoj** [comment]
Can I have some feedback on this PR please :pray:

**Comment by emilykl** [comment]
Absolutely @archmoj , thanks for the nudge 😊 

Working on a more detailed review but I have some comments on naming:

- I've never heard the term `striding` before and can't find examples of it being used with this meaning. How about `line` instead?
- `capitalize` is clear but I don't like that it's a verb while all the other font properties are nouns. How about `capitalization`, or if that's too long, then `caps`?
  - For the options, maybe `titlecase` instead of `word`? I don't feel strongly one way or the other but another option to consider
  
Also, perhaps add a mock or two for `striding='over'` and `striding='through'`.

File: src/plots/font_attributes.js
Comment (alexcjohnson) [inline src/plots/font_attributes.js:92]: I might suggest
- `'normal'` instead of `'none'`, since `'none'` could be interpreted as no caps, ie lowercase.
- `'upper'` and `'lower'` instead of `'uppercase'` and `'lowercase'` since the name `capitalize` already says we're talking about case.

File: src/plots/font_attributes.js
Comment (alexcjohnson) [inline src/plots/font_attributes.js:110]: `text-decoration` has [so many options](https://developer.mozilla.org/en-US/docs/Web/CSS/text-decoration) I kind of think we should just call the attribute `decoration`, leave it as an unstructured string, and point users to the reference so if they want `'underline overline wavy red'` they can have it.

File: src/plots/cartesian/axis_defaults.js
Comment (alexcjohnson) [inline src/plots/cartesian/axis_defaults.js:121]: Repeating my [comment from the last font PR](https://github.com/plotly/plotly.js/pull/6956#discussion_r1562517865):

> I suspect it would be simpler to use `coerceFont` AND easier to read the logic if it accepted TWO default objects - the first letting you pass on an entire unchanged font object, the second letting you override just one or two parts. If we did that, this call would turn into something like
`Lib.coerceFont(coerce, ‘hoverlabel.font’, globalHoverLabel.font, {color: globalHoverLabel.font.color || hoverBorder})`

In the case here, it would end up like:
```js
Lib.coerceFont(
    coerce,
    'title.font',
    font,
    {size: Lib.bigFont(font.size), color: dfltFontColor}
);
```

File: src/components/colorbar/defaults.js
Comment (alexcjohnson) [inline src/components/colorbar/defaults.js:133]: seems like this should be inverted now, something like:
```js
var dfltTitleFont = Lib.extendFlat({}, font, {
    family: tickFont.family,
    size: Lib.bigFont(tickFont.size)
});
```

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js:110]: Not all those options are working as SVG attribute.
Here is another link: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/text-decoration

File: src/plots/cartesian/axis_defaults.js
Comment (archmoj) [reply src/plots/cartesian/axis_defaults.js]: Thanks very much @alexcjohnson
#6956 was quite big PR and I didn't want to include refactoring changes like that.
But here it looks like the right place to have it.
Please see [150f0c8](https://github.com/plotly/plotly.js/pull/6983/commits/150f0c80295f5faf47b854a83989b7219ff1ac4f)

File: src/plots/font_attributes.js
Comment (emilykl) [reply src/plots/font_attributes.js:92]: Building on this -- why not just call the attribute `case`?

Then you could have `case: 'upper'`, `case: 'lower'`, and `case: 'title'` and `case: 'normal'`.

Regarding the last one, there's probably a better name for "just leave the casing as-is, don't change it" but can't think of one at the moment.

File: src/components/colorbar/defaults.js
Comment (archmoj) [reply src/components/colorbar/defaults.js]: Good point. Addressed in [d7c6f8f](https://github.com/plotly/plotly.js/pull/6983/commits/d7c6f8f2278961c5372f367b86e1ae29ea7155d3).
Thank you!

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js:92]: I do like `case` better than `caps`. 
In fact we do have a `caps` attribute for `isosurface` and `volume` which is something different. 
How about `case: 'caps'` instead of`case: 'title'`?

File: src/plots/font_attributes.js
Comment (emilykl) [reply src/plots/font_attributes.js:110]: I'm on the fence -- I think `decoration` is a much better API, and it's more future-proof for if/when SVG supports more decoration options.

But at the same time it's potentially frustrating for users if we name it the same as the CSS property but only support a fraction of the options. We can state clearly in the docs which properties are supported, but not everyone will read the docs carefully.

 I wonder how likely it is that SVG will support additional decoration options in the near-mid future.

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js]: Alright. I revisited the API in regards to your comments and other considerations.
For the attribute name `textcase` is declared instead of `capitalize`.
Instead of *word* option I wanted to use *headline* (or *title*) but then I thought we should not use them as `text-transform: 'capitalize'` is different in a way that it capitalizes all the words. See `capitalize` notes [here](https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform). As a results *caps* is used in the API.

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js]: This option is now named `decorline`. After lots of testing (including Chrome version used in Kaleido), I don't see we could/should expose more options now.
Alternatively we may rename this attribute to `underline` as boolean which is the most useful option (IMHO) and also allow other options as flag list string.

**Comment by archmoj** [comment]
This PR is ready for review :pray:

File: src/plots/font_attributes.js
Comment (emilykl) [reply src/plots/font_attributes.js]: I'm fine with `textcase` but I don't like `caps` -- I would expect that to mean all-uppercase. Maybe `wordcaps`? I don't love that either but I think it's clearer than `caps`.

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js]: Also another option might be to call it `lineposition` (or `linepos`)? 
This might be better as we declare the position of decoration line(s).

Comment (archmoj) [reply src/plots/font_attributes.js]: How do you feel about naming it `lined`?
```
lined: 'under'
```

Comment (emilykl) [reply src/plots/font_attributes.js]: How about `textline`?

**Review by emilykl — APPROVED** [review]
🔥 thanks @archmoj !!

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js]: `textline`, `line`, as well as `decorline` are not favorites because they may point to the lines drawn around each letter.

Comment (emilykl) [reply src/plots/font_attributes.js]: Ok, this one is tricky to name but out of the suggested options I think I like `lineposition` best so far.

Comment (archmoj) [reply src/plots/font_attributes.js]: Applied in [a344464](https://github.com/plotly/plotly.js/pull/6983/commits/a34446425f8aa9dcef40968c5fb03b0694b56d54).

**Comment by archmoj** [comment]
Thanks very much @emilykl for the review and I am also happy with the results at this point.

I would like to have additional approvals before merging. 
@gvwilson @alexcjohnson @LiamConnors 
Please review the API i.e. new attribute names and options in [src/plots/font_attributes.js](https://github.com/plotly/plotly.js/pull/6983/files#diff-0c0fc9b6c2b8dd0aaeb6514354262c42eeb1042221ff5ac05a3fab1bc9c365d1)

Thank you!

File: src/components/legend/defaults.js
Comment (gvwilson) [inline src/components/legend/defaults.js:37]: why '1.1' ?

File: src/components/colorbar/defaults.js
Comment (gvwilson) [inline src/components/colorbar/defaults.js:129]: so are individual properties for weight, style, etc. being replaced by a single `family` property?

File: src/plots/font_attributes.js
Comment (gvwilson) [inline src/plots/font_attributes.js:105]: I saw these values `'normal'`, `'word caps'`, etc. appear earlier - should they be defined somewhere as constants or is that just adding useless indirection?

File: src/plots/font_attributes.js
Comment (gvwilson) [inline src/plots/font_attributes.js:134]: nice

**Review by gvwilson — COMMENTED** [review]
I left a few small comments on the first few files, but I'm nowhere near understanding this code base well enough to say anything substantial. It looks like an extraordinary amount of work - thank you for wading through it.

File: src/components/colorbar/defaults.js
Comment (archmoj) [reply src/components/colorbar/defaults.js:129]: No. `weight`, `style` and other styling options are inherited from global layout font. 
`Family`  may be inherited from `colorbar.tickfont` if specified.
Good eye!
This was adjusted in respect to @alexcjohnson's [comment](https://github.com/plotly/plotly.js/pull/6983#discussion_r1586410831).

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js:105]: Good point.
We could declare these keys in a separate file as constant and used them here as well as in 
[src/components/drawing/index.js](https://github.com/plotly/plotly.js/pull/6983/files#diff-e8c93e83ebf0c3a486f3187d44d91b1709f51bccaba559fcadab2a3e3f9f56c6).
But it's easier to read them here and possibly provide a description if needed.

BTW I totally agree with you that to improve performance we could/should declare these constant arrays as constants in higher scope i.e. in several places in our source code (and possibly after switching to ES6).

File: src/plots/font_attributes.js
Comment (archmoj) [reply src/plots/font_attributes.js:134]: Thank you!

File: src/snapshot/tosvg.js
Comment (alexcjohnson) [inline src/snapshot/tosvg.js:133]: Should we be doing this even in interactive mode? ie whenever we would be setting one of these attrs to `'none'` we instead set it to `undefined`?

File: src/traces/parcats/defaults.js
Comment (alexcjohnson) [inline src/traces/parcats/defaults.js:107]: Do we need `autoShadowDflt` in both the `coerceFont` call and in the `attribute definition`? Can't we make it so the one in the attribute definition is enough?

File: src/components/legend/defaults.js
Comment (archmoj) [reply src/components/legend/defaults.js:37]: Good question. It's not related to changes of this PR so I need to look into the previous pull requests.
Possibly we wanted to have a larger size for the title and `1.2` was too large.

File: src/traces/parcats/defaults.js
Comment (archmoj) [reply src/traces/parcats/defaults.js:107]: By looking into the `dlft` of the attribute?
I think that would become more complex.
This is just for few places e.g. in `parcats` and `parcoords` where we already have text shadows by default.

Comment (alexcjohnson) [reply src/traces/parcats/defaults.js:107]: I bet now that we have this we're going to add it to more things... and aside from being repetitive it just feels risky to require anyone adding this to know they need to add it in both places.

File: src/components/legend/defaults.js
Comment (archmoj) [reply src/components/legend/defaults.js:37]: `1.2` was used for main legend title here: 
https://github.com/plotly/plotly.js/blob/c50363479539e943b1682c83c734c1ed4a7f8704/src/lib/index.js#L1423-L1425
So for group titles a value used between 1.2 and 1.

File: src/snapshot/tosvg.js
Comment (archmoj) [reply src/snapshot/tosvg.js]: Addressed in [9416441](https://github.com/plotly/plotly.js/pull/6983/commits/94164413ec95370ef18eff39b083884019c3e55e).

File: src/traces/parcats/defaults.js
Comment (archmoj) [reply src/traces/parcats/defaults.js:107]: I see your point there. But unfortunately changing this require notable changes to the way `coerceFont` is called or alternatively making an exception for `shadow` in `Lib.coerce`?
I'd be happy to give it a try in a separate PR as it could complicate this PR.

Comment (alexcjohnson) [reply src/traces/parcats/defaults.js:107]: Fair enough. Let’s then just make sure to address this the next time we add `autoShadowDflt` to another attribute.

**Review by alexcjohnson — APPROVED** [review]
💃 Looks great, this will make a really nice addition!

**Pull request merged**
